lib/libvfio-user.c

```diff
@@ -920,133 +894,284 @@ static int
+            sizeof(struct vfio_user_device_feature));
+
+    struct vfio_user_device_feature *res = msg->out.iov.iov_base;
+    res->argsz = msg->out.iov.iov_len;
+
+    switch (req->flags & VFIO_DEVICE_FEATURE_MASK) {
+        case VFIO_DEVICE_FEATURE_MIGRATION: {
+            struct vfio_user_device_feature_migration *mig =
+                (void *)res->data;
+            // FIXME are these always supported? Can we consider to be
+            // "supported" if said support is just an empty callback?
+            //
+            // We don't need to return RUNNING or ERROR since they are
+            // always supported.
+            mig->flags = VFIO_MIGRATION_STOP_COPY
+                        | VFIO_MIGRATION_PRE_COPY;
+            return 0;
         }
-    } else {
-        vfu_log(vfu_ctx, LOG_ERR,
-                "dirty pages: get [%#llx, %#llx): buffer too small (%u < %zu)",
-                (ull_t)range_in->iova, (ull_t)range_in->iova + range_in->size,
-                dirty_pages_in->argsz, argsz);
+
+        case VFIO_DEVICE_FEATURE_MIG_DEVICE_STATE: {
+            struct vfio_user_device_feature_mig_state *state =
+                (void *)res->data;
+            state->device_state = migration_get_state(vfu_ctx);
+            return 0;
+        }
+        
+        default: return ERROR_INT(EINVAL);
     }
+}
 
-    return 0;
+static int
+handle_migration_device_feature_set(vfu_ctx_t *vfu_ctx, uint32_t feature,
+                                    struct vfio_user_device_feature *res)
+{
+    assert(feature == VFIO_DEVICE_FEATURE_MIG_DEVICE_STATE);
+
+    struct vfio_user_device_feature_mig_state *state = (void *)res->data;
+
+    return migration_set_state(vfu_ctx, state->device_state);
 }
 
 static int
-handle_dirty_pages(vfu_ctx_t *vfu_ctx, vfu_msg_t *msg)
+handle_dma_device_feature_get(vfu_ctx_t *vfu_ctx, vfu_msg_t *msg,
+                              struct vfio_user_device_feature *req)
 {
-    struct vfio_user_dirty_pages *dirty_pages = msg->in.iov.iov_base;
-    int ret;
+    const size_t header_size = sizeof(struct vfio_user_device_feature)
+        + sizeof(struct vfio_user_device_feature_dma_logging_report);
+
+    struct vfio_user_device_feature_dma_logging_report *rep =
+        (void *)req->data;
+
+    ssize_t bitmap_size = get_bitmap_size(rep->length, rep->page_size);
+    if (bitmap_size < 0) {
+        return bitmap_size;
+    }
+
+    msg->out.iov.iov_len = header_size + bitmap_size;
+
+    if (req->argsz < msg->out.iov.iov_len) {
+        msg->out.iov.iov_len = 0;
+        return ERROR_INT(EINVAL);
+    }
+
+    msg->out.iov.iov_base = calloc(1, msg->out.iov.iov_len);
+
+    if (msg->out.iov.iov_base == NULL) {
+        return ERROR_INT(ENOMEM);
+    }
+
+    memcpy(msg->out.iov.iov_base, msg->in.iov.iov_base, header_size);
+
+    struct vfio_user_device_feature *res = msg->out.iov.iov_base;
+
+    res->argsz = msg->out.iov.iov_len;
 
+    dma_controller_t *dma = vfu_ctx->dma;
+
+    assert(dma != NULL);
```

Comment:
The server might not have initialized DMA, so this should be handled gracefully.